accelerate-llvm-ptx/src/Data/Array/Accelerate/LLVM/PTX/CodeGen/FoldSeg.hs

```diff
@@ -285,22 +288,24 @@ mkFoldSegP_warp aenv repr@(ArrayR shr tp) intTp combine mseed marr mseg = do
   dev <- liftCodeGen $ gets ptxDeviceProperties
   --
   let
-      (arrOut, paramOut)  = mutableArray repr "out"
-      (arrIn,  paramIn)   = delayedArray "in"  marr
-      (arrSeg, paramSeg)  = delayedArray "seg" mseg
-      paramEnv            = envParam aenv
+      (arrOut, paramOut) = mutableArray repr "out"
+      (arrIn,  paramIn)  = delayedArray "in"  marr
+      (arrSeg, paramSeg) = delayedArray "seg" mseg
+      paramEnv           = envParam aenv
       --
-      config              = launchConfig dev (CUDA.decWarp dev) dsmem grid gridQ
-      dsmem n             = warps * per_warp_bytes
+      config             = launchConfig dev (CUDA.decWarp dev) dsmem grid gridQ
+      dsmem n            = warps * per_warp_bytes
         where
-          warps           = (n + ws - 1) `P.quot` ws
+          warps          = (n + ws - 1) `P.quot` ws
       --
-      grid n m            = multipleOf n (m `P.quot` ws)
-      gridQ               = [|| \n m -> $$multipleOfQ n (m `P.quot` ws) ||]
+      grid n m           = multipleOf n (m `P.quot` ws)
+      gridQ              = [|| \n m -> $$multipleOfQ n (m `P.quot` ws) ||]
       --
-      per_warp_bytes      = (per_warp_elems * bytesElt tp) `P.max` (2 * bytesElt tp)
-      per_warp_elems      = ws + (ws `P.quot` 2)
-      ws                  = CUDA.warpSize dev
+      per_warp_bytes -- segment smem is aliassed with the smem to communicate values
+    --  | useShfl dev    =  4 * 2 -- this sounds right, but crashes..
```

Comment:
@tmcdonell This is the only big questionmark left, it works as is but I was fairly confident that we should be able to add this line, as we don't need more memory in the shfl version. Other than this, it's ready to merge I think

Reply:
This should have been `2 * sizeOf (undefined::Int)`; a change I missed when it was converted from `Int32`. We just never ran into it because that memory was always larger than that anyway due to it also being used for the per-warp memory.

Reply:
I discovered this after rewriting it to use a broadcast instruction instead, but oh well.

Reply:
Aah, that would explain! I was really confused at the time.